### PR TITLE
Progress non-blocking operations in OFI

### PR DESCRIPTION
**Pull request by jhh67**
Endpoints that have pending non-blocking operations must be progressed while waiting for the operations to complete. Consider non-blocking `on` statements such as those that occur as a result of the `for loc in Locales do on loc` idiom. The compiler-generated code loops through all locales, invokes a non-blocking `on` operation on each one, and issues the next `on` operation before the previous one completes. Upon completion, each `on` statement decrements an atomic counter, on which the main thread waits until the counter drops to zero. The problem is that the main thread does not progress the network stack while waiting on the atomic variable, and as a result the application may come to a halt waiting for the atomic variable to be decremented while there are `on` operations yet to be transmitted.

This PR adds the function `chpl_comm_ensure_progress` that must be called when a task is waiting on a condition outside of the comm layer (e.g., for an atomic counter to drop to zero) and that condition will be made true by a non-blocking communication operation (e.g., a non-blocking `on` statement). This function ensures that the endpoint associated with the task is progressed so that the non-blocking operation is transmitted successfully.

While implementing this PR I cleaned up the interface for non-blocking PUTs to allow them to be progressed once they are implemented properly (currently they are implemented as blocking PUTs, see the `chpl_comm_put_nb` implementation).

Note that this PR only affects the message-order-fence MCM mode using completion queues. I will fix the message-order MCM mode and completion counters in subsequent PRs to keep this one manageable. 

Resolves https://github.com/Cray/chapel-private/issues/6248 and resolves https://github.com/Cray/chapel-private/issues/6249.

Signed-off-by: John H. Hartman <jhh67@users.noreply.github.com>

### CODE REVIEW

File: modules/internal/Atomics.chpl [hunk 1]
```diff
@@ -405,6 +407,7 @@ module Atomics {
     inline proc const waitFor(val:bool, param order: memoryOrder = memoryOrder.seqCst): void {
       on this {
         while (this.read(order=memoryOrder.relaxed) != val) {
+          chpl_comm_ensure_progress();
```
Comment (jabraham17) [inline modules/internal/Atomics.chpl:410]: Does AtomicT.waitFor also need this?

File: runtime/src/comm/ofi/comm-ofi.c [hunk 2]
```diff
@@ -5317,35 +5374,32 @@ void amPutDone(c_nodeid_t node, amDone_t* pAmDone) {
...
+  // purposes of MCM conformance, and the memory effects of the AM must be
+  // visible to that initiator by the time it sees the'done' indicator.  In
+  // the delivery-complete mode everything is already visible.  In the
+  // message-order and message-order-fence modes we have to force PUT and AMO
+  // visibility.
   //
   if (mcmMode != mcmm_dlvrCmplt) {
-    forceMemFxVisAllNodes(mcmMode == mcmm_msgOrdFence /*checkPuts*/,
-                          true /*checkAmos*/,
+    forceMemFxVisAllNodes(true /*checkPuts*/, true /*checkAmos*/,
```
Comment (jabraham17) [inline runtime/src/comm/ofi/comm-ofi.c:5387]: Is there the potential for perf issues here? Since are always forcing puts?

**Review by jabraham17 — APPROVED** [review]

File: modules/internal/Atomics.chpl (hunk 1 above)
Comment (jhh67) [reply modules/internal/Atomics.chpl:410]: Yes, I know I made that change but lost it somehow. Good catch.

File: runtime/src/comm/ofi/comm-ofi.c (hunk 2 above)
Comment (jhh67) [reply runtime/src/comm/ofi/comm-ofi.c:5387]: Possibly in `message-order` MCM mode, which we don't really use, and it must be done to ensure MCM conformance. The MCM requires that all PUTs performed by an active message (AM) are visible when the active message completes. The function `amPutDone` does a PUT of a `done` variable to indicate that the AM is complete. Message ordering ensures that any pending PUTs to the same target locale will be visible before the PUT to `done` is visible; the problem is with PUTs to other locales.  For all MCM modes except for `delivery-complete` this requires an action to force visibility, which is what `forceMemFxVisAllNodes` does. Previously, it only checked PUTs when in `message-order-fence` mode, but it should do so for all MCM modes that are not `delivery-complete`.

**Pull request merged**